tests/test_expected_protocol.py

```diff
@@ -43,6 +50,41 @@ def test_simple_protocol():
         instr.simple = 4.5
 
 
+def test_not_all_communication_used():
+    """Test whether unused communication raises an error."""
+    with raises(AssertionError) as exc:
+        with expected_protocol(BasicTestInstrument,
+                               [('VOLT?', 3.14),
+                                ('VOLT 4.5 V',),
+                                ]) as instr:
+            assert instr.simple == 3.14
+    assert str(exc.value) == "Not all messages exchanged."
+
+
+def test_non_empty_write_buffer():
+    with raises(AssertionError) as exc:
+        with expected_protocol(BasicTestInstrument,
+                               [('VOLT?', 3.14),
+                                ]) as instr:
+            instr.adapter.write_bytes(b"VOLT")
+            instr.adapter._index = 1
+    assert str(exc.value) == "Non empty write buffer."
+
+
+def test_non_empty_read_buffer():
+    with raises(AssertionError) as exc:
+        with expected_protocol(BasicTestInstrument,
+                               [('VOLT?', 3.14),
+                                ]) as instr:
+            instr.write("VOLT?")
+    assert str(exc.value) == "Non empty read buffer."
+
+
+def test_preprocess():
```

Comment:
Good that we have that one, the external behaviour/usage should be like this, however we deal with this internally (if we find a better way)

Reply:
Defining the preprocess on Instrument level, fixes all problems and is a tiny change: adding it to the Instruments values method instead of the Adapter instantiation.

Requiring to define preprocess in the Instrument for tests, does not break anything (there are no tests yet). And if you write tests, you can make that change in the process.

Reply:
That's a good workaround until we resolve this properly.